cloudinit/sources/DataSourceVMware.py

```diff
@@ -0,0 +1,780 @@
+    """
+    Setting the hostname:
+        The hostname is set by way of the metadata key "local-hostname".
+
+    Setting the instance ID:
+        The instance ID may be set by way of the metadata key "instance-id".
+        However, if this value is absent then then the instance ID is
+        read from the file /sys/class/dmi/id/product_uuid.
+
+    Configuring the network:
+        The network is configured by setting the metadata key "network"
+        with a value consistent with Network Config Versions 1 or 2,
+        depending on the Linux distro's version of cloud-init:
+
+            Network Config Version 1 - http://bit.ly/cloudinit-net-conf-v1
+            Network Config Version 2 - http://bit.ly/cloudinit-net-conf-v2
+
+        For example, CentOS 7's official cloud-init package is version
+        0.7.9 and does not support Network Config Version 2. However,
+        this datasource still supports supplying Network Config Version 2
+        data as long as the Linux distro's cloud-init package is new
+        enough to parse the data.
+
+        The metadata key "network.encoding" may be used to indicate the
+        format of the metadata key "network". Valid encodings are base64
+        and gzip+base64.
+    """
+
+    dsname = "VMware"
+
+    def __init__(self, sys_cfg, distro, paths, ud_proc=None):
+        sources.DataSource.__init__(self, sys_cfg, distro, paths, ud_proc)
+
+        self.data_access_method = None
+        self.vmware_rpctool = VMWARE_RPCTOOL
+
+    def _get_data(self):
+        """
+        _get_data loads the metadata, userdata, and vendordata from one of
+        the following locations in the given order:
+
+            * guestinfo
+            * envvars
+        """
+
+        # Initialize the locally scoped metadata, userdata, and vendordata
+        # variables. They are assigned below depending on the detected data
+        # access method.
+        md, ud, vd = None, None, None
+
+        # First check to see if there is data via guestinfo.
+        if self.vmware_rpctool:
+            md = guestinfo("metadata", self.vmware_rpctool)
+            ud = guestinfo("userdata", self.vmware_rpctool)
+            vd = guestinfo("vendordata", self.vmware_rpctool)
+
+            if md or ud or vd:
+                self.data_access_method = DATA_ACCESS_METHOD_GUESTINFO
+
+        # If data has not been loaded, check via envvars.
+        if not self.data_access_method:
+            if os.environ.get(VMX_GUESTINFO, ""):
+                md = guestinfo_envvar("metadata")
+                ud = guestinfo_envvar("userdata")
+                vd = guestinfo_envvar("vendordata")
+
+                if md or ud or vd:
+                    self.data_access_method = DATA_ACCESS_METHOD_ENVVAR
+
+        if not self.data_access_method:
+            LOG.error("failed to find a valid data access method")
+            return False
+
+        LOG.info("using data access method %s", self.data_access_method)
```

Comment:
This is good information that should probably also be returned by a  DatasourceVMware._get_subplatform method to better describe to instance-data.json what "sub flavor" of metadata sources we have detected on the system. See [DataSourceOVF](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceOVF.py#L400-L404) and [DataSourceNoCloud](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceNoCloud.py#L204-L210) for examples of the types of descriptive strings that might allow someone to better triage where to datasource discovered metadata came from.

Reply:
Do you think we should implement `_get_subplatform` or just change `self.data_access_method` to `self.metadata_address`? It looks like the base DS will use that attribute for the sub platform if it exists.

Reply:
Eh, I just did what you suggested as it was cleaner within the existing implementation.